File: http-server-netty/src/test/groovy/io/micronaut/docs/http/server/exception/aterrorexceptionhandlerprecedence/BookController.java
package io.micronaut.docs.http.server.exception.aterrorexceptionhandlerprecedence;

import io.micronaut.context.annotation.Requires;
import io.micronaut.http.MediaType;
import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.micronaut.http.annotation.Produces;

@Requires(property = "spec.name", value = "AtErrorExceptionHandlerPrecedenceSpec")
//tag::clazz[]
@Controller("/books")
public class BookController {
    @Produces(MediaType.TEXT_PLAIN)
    @Get("/stock/{isbn}")
    Integer stock(String isbn) {
        throw new OutOfStockException();
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/docs/placeholders/BookController.py
from micronaut.context.annotation import Requires
from micronaut.http import MediaType
from micronaut.http.annotation import Controller, Get


@Requires(property="spec.name", value="DocumentedPlaceholderSpec")
@Controller("/placeholders/books")
class BookController:
    @Get("/{id}", produces=MediaType.TEXT_PLAIN)
    def find(self, id: int) -> str:
        return f"Book {id}"
